### PR TITLE
Refactor for plugin capability, add pods and npm clean commands

### PR DESCRIPTION
**Pull request by mikehardy**
Hi there!

A few of us in the react-native-community were discussing how we could never escape the need to completely wipe out state and start clean, and the react-native command line should just have a command for it.

The CLI maintainers mentioned we could probably do it as a plugin pretty quickly, and I found this project, and thought you might be okay with refactoring it so it could be used as a plugin?

So I did that, and I also added a couple extra possible commands - pod delete, pod update, and npm cache clean

It should work with both CLI 1.x and the upcoming 2.x for react-native 0.60, and you just need to do `react-native clean-project` or `react-native clean-project-auto` to have it either take you through the questions, or just purge everything respectively

If it's merge-worthy, it fixes https://github.com/react-native-community/cli/issues/387

### CODE REVIEW

**Comment by pmadruga** [comment]
Thanks for this and yes, it can totally be used as a plugin. I will only ask for retroactive compatibility and any changes requested that may have come from the review. Jumping to the review now.

File: package.json [hunk 1]
```diff
@@ -8,6 +8,8 @@
   "main": "./source/index.js",
   "bin": "./source/index.js",
   "scripts": {
+    "create-example": "react-native init example && cd example && yarn add -D react-native-clean-project",
```
Comment (pmadruga) [inline package.json]: Just curious on why is this necessary?

File: source/plugin.js [hunk 2]
```diff
@@ -0,0 +1,27 @@
+module.exports = [
+  {
+    name: 'clean-project-auto',
```
Comment (pmadruga) [inline source/plugin.js]: The explanation for the why of these two options is necessary in the README.

File: source/internals.js [hunk 3]
```diff
@@ -0,0 +1,242 @@
+const { spawn } = require('child_process');
```
Comment (pmadruga) [inline source/internals.js]: This file became rather big, I would rather create an 'internals' folder with the tasksList, the ask* and the gets into its own files.

**Comment by mikehardy** [comment]
Oh great - I'm sure we can get this in a shape you like then. I'm pretty agnostic on everything *except* that things should work and be easy to maintain :-). I'll go through your review now

File: source/plugin.js (hunk 2 above)
Comment (mikehardy) [reply source/plugin.js]: Totally agreed - I just wanted to see if you receptive before I did the administrivia

File: source/internals.js (hunk 3 above)
Comment (mikehardy) [reply source/internals.js]: I would agree with that as well - it's pretty cluttered right now, but again I wanted to do something more minimal, like a Proposal Request more, first, since you're receptive I'll take your advice and re-shape this

**Comment by mikehardy** [comment]
Worth mentioning that I agree 100% on backwards compatibility, I see no reason this should be a breaking change in any way, it's really just making the code so it can be wrapped, then doing so. I think the community will really like this - they were excited when I showed them this package, since the UI and feedback is really cool for users

File: package.json (hunk 1 above)
Comment (mikehardy) [reply package.json]: this was necessary because I saw a test entry in there but couldn't see how you were doing tests in here, so I went with ad-hoc testing using a quick example app I could sync changes into and run. Not automated but it is a very quick test/debug cycle, I must have run it 50 times while slowly making changes.

You mention tapejs, would you be willing to put an example of how you want to test things into master, then I can copy that pattern and exercise the changes?

Or if it is already there and I'm mixing something obvious just clue me in and I'll get to it

File: source/plugin.js (hunk 2 above)
Comment (thymikee) [reply source/plugin.js]: please add a `description` to the tasks for more helpful `--help` output of the CLI

File: package.json (hunk 1 above)
Comment (pmadruga) [reply package.json]: You're right, there isn't a setup to run tests yet (sorry for that). Let me setup it and put it in master. I'll go with jest instead. 

I'll write here as soon as it's done.

Comment (pmadruga) [reply package.json]: @mikehardy setup for running tests and a small test is now on master.

File: source/plugin.js (hunk 2 above)
Comment (mikehardy) [reply source/plugin.js]: fantastic - I'll get to it, thanks!

File: package.json (hunk 1 above)
Comment (mikehardy) [reply package.json]: Fantastic, I'll get after it today

**Comment by mikehardy** [comment]
Okay @pmadruga (and @thymikee if you have interest) - I think this one is an actual Pull Request as opposed to a quick-and-dirty Proposal Request like it was originally. Please take a fresh look and let me know what think?

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -7,14 +7,14 @@ Please note we have a code of conduct, please follow it in all your interactions
...
-    Pull Request would represent. The versioning scheme we use is [SemVer](http://semver.org/).
-4.  You may merge the Pull Request in once you have the sign-off of two other developers, or if you
-    do not have permission to do that, you may request the second reviewer to merge it for you.
+1. Ensure any install or build dependencies are removed before the end of the layer when doing a
+   build.
+1. Update the README.md with details of changes to the interface, this includes new environment
+   variables, exposed ports, useful file locations and container parameters.
+1. Increase the version numbers in any examples files and the README.md to the new version that this
+   Pull Request would represent. The versioning scheme we use is [SemVer](http://semver.org/).
+1. You may merge the Pull Request in once you have the sign-off of two other developers, or if you
```
Comment (mikehardy) [inline CONTRIBUTING.md]: All the changes in CONTRIBUTING were just markdown lint

File: package.json [hunk 5]
```diff
@@ -8,6 +8,8 @@
   "main": "./source/index.js",
   "bin": "./source/index.js",
   "scripts": {
+    "dev-init": "npx react-native init example",
+    "dev-sync": "cp -rv package.json source react-native.config.js example/node_modules/react-native-clean-project/",
```
Comment (mikehardy) [inline package.json]: Even with the automated testing, I took tremendous comfort in having quick access to a rebuildable example project with quick install of local module changes, so I left the example items in (here, .gitignore, and ignored in jest config). If you feel strongly that you don't like them, I can drop them

File: package.json [hunk 6]
```diff
@@ -27,7 +29,10 @@
   },
   "homepage": "https://github.com/pmadruga/react-native-clean-project#readme",
   "devDependencies": {
-    "eslint": "5.2.0",
+    "eslint": "^5.16.0",
     "jest": "^24.8.0"
+  },
+  "rnpm": {
+    "plugin": "./source/plugin.js"
```
Comment (pmadruga) [inline package.json]: I guess it's normal boilerplate to have the './source/plugin' reference both here and in the react-native.config.js. Should they always match?

File: source/plugin.js [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+    func: () => {
+
+      console.log('');
+      console.log('Executing fully-automated project clean.');
+      console.log('Use \'react-native clean-project\' for more control');
+      console.log('Use \`./node_modules/.bin/react-native-clean-project\' for total control');
+      console.log('');
+
+      rlInterface.close(); // if we don't do this it hangs waiting for input
```
Comment (pmadruga) [inline source/plugin.js]: The tests' results also raise a warning regarding an open task that's keeping preventing jest from exiting, and it's related to `rlInterface`. You can check it by running `./node_modules/.bin/jest --detectOpenHandles`. Can you give it a look, please?

File: CONTRIBUTING.md (hunk 4 above)
Comment (mikehardy) [reply CONTRIBUTING.md]: Dang - must have to do with spacing / newlines. In markdown for numbered lists just repeat 1. and it works, assuming you do it right :-). I was just linting anyway - no good deed unpunished. I'll fix it

File: CONTRIBUTING.md [hunk 8]
```diff
@@ -119,7 +119,7 @@ that are not aligned to this Code of Conduct, or to ban temporarily or
 permanently any contributor for other behaviors that they deem inappropriate,
 threatening, offensive, or harmful.
 
-### Scope
+### Code of Conduct Scope
```
[the remark this answers is not in the document]
Comment (mikehardy) [reply CONTRIBUTING.md]: This was markdown lint - there were two headings with identical contents, so generated header anchors won't work. So I just changed one slightly. It's just lint so I don't really care, but it's a valid technical change - if it's a poor stylistic change, easy to revert. What do you think?

File: source/index.js [hunk 9]
```diff
@@ -1,204 +1,45 @@
...
-
-askiOS()
-  .then(askAndroid)
-  .then(askNodeModules)
-  .then(askBrew)
+const options = require('./internals/options');
+const { executeTask } = require('./internals/executor');
+const { tasks } = require('./internals/tasks');
+
+options.askiOS()
```
[the remark this answers is not in the document]
Comment (mikehardy) [reply source/index.js]: Okay - let me think about that one (to do it well) and address it

File: package.json (hunk 6 above)
Comment (mikehardy) [reply package.json]: Yes - one is for react-native-community/cli v1 (the rnpm config) and one is for v2 (where they migrate to react-native.config.js. I wanted it to work now and in the pending release so I included the right hooks for both. I don't believe we can do comments in JSON or I'd leave a comment to remove the rnpm implementation later (once CLI v1 is long gone) but that would be the idea

File: source/plugin.js (hunk 7 above)
Comment (mikehardy) [reply source/plugin.js]: I gave that one a long look and I'm not sure how to fix it well. My general take was that it was like that before - by which I mean structurally you have the readline opened immediately whether it will be used or not, then you close it in an unrelated area. My conversion to a plugin style didn't relate to that though it did mean I had to be careful to close it myself in plugin.js. Unfortunately even though I'm closing it there it isn't closing when run by jest.

It was late enough at night and I'd worked enough on it that after contemplating the "no fix" path (no real consequence) I let it go. It's an existing issue with the code that needs cleaning but it'll require re-organizing how readline is used in the options.js which are orthogonal to my change.

Can we let it go for this PR?

**Comment by mikehardy** [comment]
Thanks so much for the quick review, great comments and I'll fix what's obviously warm while waiting for your reply on the rest.

File: source/index.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (mikehardy) [reply source/index.js]: Hmm - on review I think this is original structure as well - they were mixed to start - and is orthogonal to my changes. I need to draw boundaries or I'll be on this forever. If it was simple I'd go after it immediately but really it should be two phases fully separate / one for all asking, and one for all execution. It's sort of like that now but if you want to disentangle it's really going to be up to your taste and there is a large chance I won't do it the way you'll want to live with. Additionally there already were OS-specific tasks, but they weren't handled that way to begin with (brew is macOS only, pods are new and macOS only so I just followed the pattern of brew, watchman isn't on linux but is on windows and mac, there's no "choco" but that is recommended for windows etc, also some people use yarn some use npm, the package install should check for yarn.lock or package-lock.json). So there's lots of scope here to do it really right, but I followed existing pattern. I'd like to leave this for now if possible?

**Comment by mikehardy** [comment]
Okay - I've addressed all the comments, even if for the 2 more substantial ones I was essentially pleading to keep them separate. Let me know what you think - thanks!

File: CONTRIBUTING.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (pmadruga) [reply CONTRIBUTING.md]: Nop, looks good. Thank you!

File: source/index.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (pmadruga) [reply source/index.js]: You're right and I totally agree to leave it for now. I'll get to it in the future then.

(I hope my comment before didn't sound as criticism)

File: source/plugin.js (hunk 7 above)
Comment (pmadruga) [reply source/plugin.js]: You also have a point here and I'll take it in another PR. Thank you for explaining it thoroughly.

**Comment by pmadruga** [comment]
Thanks for the replies, @mikehardy . The PR looks great and exposed some of the issues that I need to address. Thumbs up from me. I'll merge as soon as possible - would just like @thymikee 's thumbs up. 

And overall, thanks for this PR.

File: source/plugin.js [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+
+      let taskPromises = [];
+      autoTasks.forEach(task => {
+        taskPromises.push(executeTask(task));
+      });
+
+      Promise.all(taskPromises).then(() => {
```
Comment (thymikee) [inline source/plugin.js]:
```suggestion
      return Promise.all(taskPromises).then(() => {
```

By returning the result from the `func` function CLI is able to intercept the error messages and know when the command succeeded/failed

File: source/index.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (mikehardy) [reply source/index.js]: :-), no worries - I love a good engineering debate, it makes great software. Just running out of time and wanted to fix the scope out of self-interest haha

**Comment by mikehardy** [comment]
Fantastic! I'm not part of the CLI team (I just run around making changes on things that drive me crazy personally) but I am going to use the heck out of this thing now. If the CLI folks want to keep this out of core I could see this plugin being the de facto standard at which point you and your module will be #InternetFamous, which is fun :-). Cheers

**Comment by mikehardy** [comment]
@thymikee I'm so glad you looked at this, I would have never known that helped the CLI, not knowing the internals. I committed your suggestion exactly as is

**Comment by mikehardy** [comment]
@pmadruga Any chance of a release with this soon :pray: :smiley: - would love to recommend this as a simple npm install vs using a git master dependency

**Comment by pmadruga** [comment]
Done :)
